src/Renci.SshNet/Security/Cryptography/HMAC.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Security.Cryptography;
+
+namespace Renci.SshNet.Security.Cryptography
+{
+    /// <summary>
+    /// Represents the info for Message Authentication Code (MAC).
+    /// </summary>
+    public sealed class HMAC : IDisposable
```

Comment:
The new type seems unnecessary. Can we just change `IKeyExchange.Create[..]Hash` to `Create[..]Hash(out bool isEncryptThenMac)` or to return a tuple `(HashAlgorithm, bool)`?

Reply:
My first revision (not committed) is `Create[...]Hash(out bool etm)`. However, compared with other argumentless `Create[...]()` methods in `IKeyExchange` interface, it looks a little bit different.

Then I took a look at other implementations, like [OpenSSH](https://github.com/openssh/openssh-portable) and [sshj](https://github.com/hierynomus/sshj). They both have dedicated data structure for MAC. Their implementation defines the detailed hash related methods plus `etm` or `isEtm` property.

https://github.com/openssh/openssh-portable/blob/master/mac.c
https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/transport/mac/MAC.java

Reply:
Right, but we already have HashInfo, which is practically the same thing. We don't need another type representing the same information.

Reply:
From my understanding, the `HashInfo` is kind of the factory. The proposed `HMAC` is the _product_. I couldn't find a way to merge them.

Reply:
How about removing the HMAC class and move ETM information to HashInfo?
You've already added the property here:
https://github.com/sshnet/SSH.NET/pull/1316/files#diff-8ad50217d0fd4634eedc20ae503cfa6584eb26b59c5922c807b581c2455f4e25R25-R28

This would protect us from the breaking change from my second comment.

Reply:
That's a good spot! The `ETM` property was added previously and then forgot to delete.
At the very beginning, I had the very similar thoughts with you which is to leverage `HashInfo` to store `ETM` Boolean value. However, later on I found the `HashInfo` is kind of factory. The `HashAlgorithm` is the _product_.

See https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Security/IKeyExchange.cs#L72 and https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Security/IKeyExchange.cs#L80

The `HashAlgorithm` is a .NET build-in class which cannot add new properties. The only way is to change `IKeyExchange.Create[..]Hash` to `Create[..]Hash(out bool isEncryptThenMac)` (which is my first revision) or to return a tuple e.g. `(HashAlgorithm, bool) Create[..]Hash()` just as @Rob-Hague pointed. I've already explained the reason why I abandoned my first revision and created dedicated `HMAC` class. See my previous comments.

I know it is a breaking change, but I don't think it will bring critical impact. There was similar successful breaking change previously. See https://github.com/sshnet/SSH.NET/commit/24838e6173e2a7d4cbb498ce5edc3e32b254ca11

Reply:
I do think we should either merge the classes or go with the `out bool` parameter in `IKeyExchange`, and I would suggest doing the latter. It is a much simpler change and creates one small break in the internals of the library where it matters a lot less.

Reply:
The `Cipher` returned from `CreateServerCipher` and `CreateClientCipher` is defined in SshNet. The `Compressor` returned from `CreateCompressor` and `CreateDecompressor` is also defined in SshNet.
The `HashAlgorithm` returned from `CreateServerHash` and `CreateClientHash` is a BCL class which prevents us from extending it.
There's already a bunch of `HMAC[...]` classes in SshNet, which has the same name with BCL `HMAC[...]`, specifically used for SSH purpose.